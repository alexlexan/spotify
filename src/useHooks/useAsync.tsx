import { useState, useEffect, useCallback, useRef } from "react";

type PromiseFn<T> = (...args: any) => Promise<T>;
export type Status = "idle" | "pending" | "success" | "error";

function useAsync<D, E, F extends PromiseFn<D>>(
  asyncFunction: F,
  token: string,
  immediate: boolean = true
) {
  const [status, setStatus] = useState<Status>("idle");
  const [value, setValue] = useState<D | null>(null);
  const [error, setError] = useState<E | null>(null);
  let mounted = useRef(true);

  const execute = useCallback(() => {
    setStatus("pending");
    setValue(null);
    setError(null);

    return asyncFunction(token)
      .then((response) => {
        mounted.current && setValue(response);
        setStatus("success");
      })
      .catch((error) => {
        setError(error);
        setStatus("error");
      });
  }, [asyncFunction, token]);

  useEffect(() => {
    if (immediate) {
      execute();
    }
    return () => {
      mounted.current = false;
    };
  }, [execute, immediate]);

  return { execute, status, value, error };
}

export default useAsync;
