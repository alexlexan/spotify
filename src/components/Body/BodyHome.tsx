import React from "react";
import Header from "../Header/Header";
import Playlists from "./Playlists/Playlists";
import PlaylistsTrack from "./Playlists/PlaylistsTrack";
import { useSelector } from "react-redux";
import useAsync from "../../useHooks/useAsync";
import CircularProgress from "@material-ui/core/CircularProgress";
import "./Body";
import Error from "../Error";
import { AppState } from "../../store/rootReducer";
import {
  PlaylistsTrackType,
  PlaylistsType,
  PlaylistsNewReleasesType,
} from "../../store/reducer";

const BodyHome = () => {
  const token = useSelector((state: AppState) => state.spotify.token);
  const { status, value, error } = useAsync<
    [PlaylistsType, PlaylistsTrackType, PlaylistsNewReleasesType],
    Error,
    getHomePlaylistsType
  >(getHomePlaylists, token);

  if (status === "error") {
    return <Error error={error} />;
  }

  return (
    <>
      <Header />
      {value && status === "success" ? (
        <>
          <Playlists playlists={value[0]} title={`Top Artists`} />
          <Playlists playlists={value[2]?.albums} title={`New Releases`} />
          <PlaylistsTrack playlists={value[1]} title={`Top Tracks`} />
        </>
      ) : (
        <div className="loader">
          <CircularProgress color="inherit" />
        </div>
      )}
    </>
  );
};

export default BodyHome;

async function getHomePlaylists<
  PlaylistsType,
  PlaylistsTrackType,
  PlaylistsNewReleasesType
>(
  token: string
): Promise<[PlaylistsType, PlaylistsTrackType, PlaylistsNewReleasesType]> {
  const getData = async (url: string, token: string) => {
    const request = new Request(url, {
      headers: {
        Authorization: "Bearer " + token,
        Accept: "application/json",
      },
    });

    const responce = await fetch(request);
    const result = await responce.json();
    return result;
  };

  const topArtists = await getData(
    `https://api.spotify.com/v1/me/top/artists?limit=5`,
    token
  );
  const topTracks = await getData(
    `https://api.spotify.com/v1/me/top/tracks?limit=10`,
    token
  );
  const newRealeses = await getData(
    `https://api.spotify.com/v1/browse/new-releases?country=SE&limit=5`,
    token
  );
  return Promise.all([topArtists, topTracks, newRealeses]);
}

type getHomePlaylistsType = typeof getHomePlaylists;
