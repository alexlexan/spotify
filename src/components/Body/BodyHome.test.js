import React from "react";
import BodyHome from "./BodyHome";
import { getHomePlaylists } from "./BodyHome";
import useAsync from "../../useHooks/useAsync";

jest.mock("../../useHooks/useAsync");

jest.mock("react-redux", () => ({
  useSelector: (selector) => selector({ spotify: { token: "string" } }),
}));

describe("test BodyHome component", () => {
  let component;

  it("should render Playlists and PlaylistsTrack components if we have success status", () => {
    useAsync.mockReturnValue({
      status: "success",
      value: ["PlaylistsType", "PlaylistsTrackType", { albums: "albums" }],
      error: null,
    });

    component = shallow(<BodyHome />);
    expect(component).toMatchSnapshot();
  });
  it("should render Loader component if we have pending status", () => {
    useAsync.mockReturnValue({
      status: "pending",
      value: ["PlaylistsType", "PlaylistsTrackType", { albums: "albums" }],
      error: null,
    });

    component = shallow(<BodyHome />);
    expect(component).toMatchSnapshot();
  });
  it("should render Error component if we have error status", () => {
    useAsync.mockReturnValue({
      status: "error",
      value: null,
      error: "error",
    });

    component = shallow(<BodyHome />);
    expect(component).toMatchSnapshot();
  });
});

describe("getHomePlaylists function test", () => {
  let fetch;
  beforeEach(() => {
    fetch = jest.spyOn(window, "fetch");
  });
  afterEach(() => {
    fetch.mockClear();
  });

  it("should return corret result", async () => {
    fetch
      .mockResolvedValueOnce({
        ok: true,
        json: async () => "topArtists",
      })
      .mockResolvedValueOnce({
        ok: true,
        json: async () => "topTracks",
      })
      .mockResolvedValueOnce({
        ok: true,
        json: async () => "newRealeses",
      });
    const result = await getHomePlaylists("token");
    expect(result).toEqual(["topArtists", "topTracks", "newRealeses"]);
    expect(fetch).toHaveBeenCalledTimes(3);
  });
});
