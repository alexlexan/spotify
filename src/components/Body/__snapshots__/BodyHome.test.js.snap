// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test BodyHome component should render Error component if we have error status 1`] = `
<Error
  error="error"
/>
`;

exports[`test BodyHome component should render Loader component if we have pending status 1`] = `
<Fragment>
  <Memo(Header) />
  <div
    className="loader"
  >
    <WithStyles(ForwardRef(CircularProgress))
      color="inherit"
    />
  </div>
</Fragment>
`;

exports[`test BodyHome component should render Playlists and PlaylistsTrack components if we have success status 1`] = `
<Fragment>
  <Memo(Header) />
  <Memo(Playlists)
    playlists="PlaylistsType"
    title="Top Artists"
  />
  <Memo(Playlists)
    playlists="albums"
    title="New Releases"
  />
  <PlaylistsTrack
    playlists="PlaylistsTrackType"
    title="Top Tracks"
  />
</Fragment>
`;
